Implement `no-route-action` rule (#2030)
ember-template-lint/ember-template-lint: docs/rule/no-route-action.md

# no-route-action

This rule disallows the usage of `route-action`.

> [ember-route-action-helper](https://github.com/DockYard/ember-route-action-helper) is a popular addon that it's only purpose was to avoid creating controllers, something we no longer have a desire to do given the current state of the framework.

Reviewer: ```suggestion
[ember-route-action-helper](https://github.com/DockYard/ember-route-action-helper) was a popular addon used to avoid creating controllers just to add actions to routes. Given the changes in Ember since ember-route-action-helper was useful, controllers are now a useful pattern we want to encourage and discourage the use of route-action.
```

Most route actions should either be sent to the controller first or encapsulated within a downstream component instead. We should never be escaping the DDAU hierarchy to lob actions up to the route.

## Examples

This rule **forbids** the following:

```hbs
{{custom-component onUpdate=(route-action 'updateFoo')}}
```

```hbs
{{custom-component onUpdate=(route-action 'updateFoo' 'bar')}}
```

> 

Reviewer: I like the pattern eslint uses to show good `Examples of incorrect code for this rule` and `Examples of correct code for this rule`.  I suggest using that pattern here.  

Incorrect:
It would be useful to see an example where someone defined an action on the route file and used route-action in the hbs file.

Correct:
Show a controller file, a route file (with no action), and the hbs file.

[example](https://eslint.org/docs/rules/no-empty)

Author: Good suggestion.

I didn't want to mix too much JS with this template lint rule. I just added a small JS followed by what is forbidden and allowed. On top of that I added a small migration example to show the whole example.

```hbs

> <CustomComponent onUpdate={{route-action 'foo'}} />

Reviewer 2: ```suggestion
<CustomComponent @onUpdate={{route-action 'foo'}} />
```

Reviewer 2: This is the case for every `<CustomComponent />` occurrence here and in the test file.

```

```hbs
<CustomComponent onUpdate={{route-action 'foo' 'bar'}} />
```

Instead, use controller actions as the following:

```hbs

> {{custom-component onUpdate=(this.updateFoo)}}

Reviewer 2: ```suggestion
{{custom-component onUpdate=this.updateFoo}}
```

This rule disallows the usage of `route-action`.

> [ember-route-action-helper](https://github.com/DockYard/ember-route-action-helper) was a popular addon used to avoid creating controllers just to add actions to routes. Given the changes in Ember since ember-route-action-helper was useful, controllers are now a useful pattern we want to encourage and discourage the use of route-action.

Reviewer: was a popular addon used to avoid creating controllers just to add actions to routes. Given the changes in Ember since ember-route-action-helper was useful, controllers are now a useful pattern we want to encourage and discourage the use of route-action.
